### ISSUE TITLE
UB in `AlignedVec` due to potential null pointer input to `NonNull::new_unchecked`

### ISSUE THREAD

**Issue by robinhundt** [rkyv/rkyv#278]
I was looking at the source code of the [`AlignedVec`](https://github.com/rkyv/rkyv/blob/master/rkyv/src/util/aligned_vec.rs) struct, when I noticed multiple places of UB because of unchecked usage of [`NonNull::new_unchecked`](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.new_unchecked). The docs clearly state that the passed pointer must be non-null. However, in the implementation of `AlignedVec`, there are multiple places ([`with_capacity`](https://github.com/rkyv/rkyv/blob/master/rkyv/src/util/aligned_vec.rs#L87-L95), [`change_capacity`](https://github.com/rkyv/rkyv/blob/master/rkyv/src/util/aligned_vec.rs#L129-L131) and [`reserve`](https://github.com/rkyv/rkyv/blob/master/rkyv/src/util/aligned_vec.rs#L300-L313)) where the returned pointers of the `alloc` API (like [`realloc`](https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#tymethod.alloc)) are directly passed to `NonNull::new_unchecked`. As the re-/allocation methods can return a null pointer due to different reasons, bypassing the null check is not sound.

One of the problematic methods:
```rust
  pub fn with_capacity(capacity: usize) -> Self {
      if capacity == 0 {
          Self::new()
      } else {
          unsafe {
              let ptr = alloc::alloc(alloc::Layout::from_size_align_unchecked(
                  capacity,
                  Self::ALIGNMENT,
              ));
              Self {
                  ptr: NonNull::new_unchecked(ptr),
                  cap: capacity,
                  len: 0,
              }
          }
      }
  }
```

Best,
Robin

**Comment by djkoloski** [comment]
Thanks for reporting this issue, I'll do a pass on `AlignedVec` and any other places that use the global allocator.

**Comment by djkoloski** [comment]
Fixed by 5b929039f46b5c69607b2861fb3d9ef540a97564
